nixos/modules/security/acme/default.nix

```diff
@@ -15,34 +14,45 @@ let
   mkAccountHash = acmeServer: data: mkHash "${toString acmeServer} ${data.keyType} ${data.email}";
   accountDirRoot = "/var/lib/acme/.lego/accounts/";
 
-  lockdir = "/run/acme/";
-  concurrencyLockfiles = map (n: "${toString n}.lock") (lib.range 1 cfg.maxConcurrentRenewals);
-  # Assign elements of `baseList` to each element of `needAssignmentList`, until the latter is exhausted.
-  # returns: [{fst = "element of baseList"; snd = "element of needAssignmentList"}]
-  roundRobinAssign = baseList: needAssignmentList:
-    if baseList == [] then []
-    else _rrCycler baseList baseList needAssignmentList;
-  _rrCycler = with builtins; origBaseList: workingBaseList: needAssignmentList:
-    if (workingBaseList == [] || needAssignmentList == [])
-    then []
-    else
-      [{ fst = head workingBaseList; snd = head needAssignmentList;}] ++
-      _rrCycler origBaseList (if (tail workingBaseList == []) then origBaseList else tail workingBaseList) (tail needAssignmentList);
-  attrsToList = lib.mapAttrsToList (attrname: attrval: {name = attrname; value = attrval;});
-  # for an AttrSet `funcsAttrs` having functions as values, apply single arguments from
-  # `argsList` to them in a round-robin manner.
-  # Returns an attribute set with the applied functions as values.
-  roundRobinApplyAttrs = funcsAttrs: argsList: lib.listToAttrs (map (x: {inherit (x.snd) name; value = x.snd.value x.fst;}) (roundRobinAssign argsList (attrsToList funcsAttrs)));
-  wrapInFlock = lockfilePath: script:
-    # explainer: https://stackoverflow.com/a/60896531
-    ''
-      exec {LOCKFD}> ${lockfilePath}
-      echo "Waiting to acquire lock ${lockfilePath}"
-      ${pkgs.flock}/bin/flock ''${LOCKFD} || exit 1
-      echo "Acquired lock ${lockfilePath}"
-    ''
-    + script + "\n"
-    + ''echo "Releasing lock ${lockfilePath}"  # only released after process exit'';
+  writeMaxConcurrentShellScript = name: text:
+    let script = pkgs.writeShellScript name text; in
+    if cfg.maxConcurrentRenewals == 0
+    then script
+    else "exec ${lib.escapeShellArg lockAndRun} ${lib.escapeShellArg script}";
```

Comment:
Instead of generating a new script, would it be more practical to inline the lock acquisition logic where it is needed? Something like `lib.optionalString cfg.maxConcurrentRenewals == 0 ....`.

That way you don't need to worry about the debug args/pipefail also.

Reply:
I don't really like merging scripts when not everything fits on the screen at once.  
It's easy to mess up existing variables or any other global state, be incompatible with shell options, etc.

We can't use a subshell here either cause it'll close the fd/lock on exit (and it does the solve everything anyways).

Reply:
A subshell may not work but a forking background process in a function does. Would you settle for moving the locking code to a bash function? It can either be inlined or "source"d into the scripts further down. The function can also exit early if `maxConcurrentRenewals == 0`, removing the need for extra conditional logic in the Nix portion. Heck if you were being really fancy, you could make it use environment variables instead of Nix templates, and thus the lock script/library could be static across certs + rebuilds.

The locking code may be off screen at this point, but the actual call to this locking function won't be, which I think fits both our criteria for simplified script generation and readability.

Reply:
I think my suggestion is a little confusing so I [made a gist](https://gist.github.com/m1cr0man/2aec78ce887c7eb56fb2e2d13773cd7e). Note I inlined the Nix stuff/used env vars just so I could test it on the CLI easier.